text/0000-rewards.md

```diff
@@ -0,0 +1,298 @@
+
+### Observation
+
+We track this data for each candidate during the approvals process:
+```
+/// Our subjective record of out availability transfers for this candidate.
+CandidateRewards {
+    /// Anyone who backed this parablock
+    backers: [AuthorityId; NumBackers],
+    /// Anyone to whome we think no-showed, even only briefly.
+    noshows: HashSet<AuthorityId>,
+    /// Anyone who sent us chunks for this candidate
+    downloaded_from: HashMap<AuthorityId,u16>,    
+    /// Anyone to whome we sent chunks for this candidate
+    uploaded_to: HashMap<AuthorityId,u16>,
+}
+```
+We no longer require this data during disputes.  
+<!-- You could optionally track a `downloaded_one: Option<AuthorityBitField>` too, for the nodes from whome we douwnloaded only one chunk, but this seems like premature optimization -->
+
+After we approve a relay chain block, then we collect all its `CandidateRewards` into an `ApprovalsTally`, with one `ApprovalTallyLine` for each validator.  In this, we compute `approval_usages` from the final run of the approvals loop, plus `0.8` for each backer.
+```
+/// Our subjective record of what we used from, and provided to, all other validators on the finalized chain
+pub struct ApprovalsTally(Vec<ApprovalTallyLine>);
+
+/// Our subjective record of what we used from, and provided to, all one other validators on the finalized chain
+pub struct ApprovalTallyLine {
+    /// Approvals by this validator which our approvals gadget used in marking candidates approved.
+    approval_usages: u32,
+    /// How many times we think this validator no-showed, even only briefly.
+    noshows: u32
+    /// Availability chunks we downloaded from this validator for our approval checks we used.
+    used_downloads: u32,
+    /// Availability chunks we uploaded to this validator which whose approval checks we used.
+    used_uploads: u32,
+}
+```
+At finality we sum these `ApprovalsTally` for one for the whole epoch so far, into another `ApprovalsTally`.  We can optionally sum them earlier at chain heads, but this requires mutablity.
+
+### Messages
+
+After the epoch is finalized, we share the first two lines of its `ApprovalTally`.
+```
+/// Our subjective record of what we used from some other validator on the finalized chain
+pub struct ApprovalTallyMessageLine {
+    /// Approvals by this validator which our approvals gadget used in marking candidates approved.
+    approval_usages: u32,
+    /// How many times we think this validator no-showed, even only briefly.
+    noshows: u32
+    /// Availability chunks we downloaded from this validator for our approval checks we used.
+    used_downloads: u32,
+}
+
+/// Our subjective record of what we used from all other validators on the finalized chain
+pub struct ApprovalsTallyMessage(Vec<ApprovalTallyMessageLine>);
+```
+
+Actual `ApprovalsTallyMessage`s sent over the wire must be signed of course, likely by the grandpa ed25519 key.
```

Comment:
Hi! I have some questions about this message exchanging step.

- Regarding the messages, they should be shared using grid topology? Using an existent protocol or a new one? 
- Will exist some timeout to receive some amount X of messages from other validators? What happens if we don't receive the messages in time?